srml/support/procedural/src/storage/impls.rs

```diff
@@ -0,0 +1,414 @@
+							res.1.previous = Some(key);
+							storage.put(&*head_key, &res);
+						}
+						// update to current head
+						Self::write_head(storage, Some(&key));
+						// return linkage with pointer to previous head
+						let mut linkage = Self::default();
+						linkage.next = Some(head);
+						linkage
+					} else {
+						// we are first - update the head and produce empty linkage
+						Self::write_head(storage, Some(&key));
+						Self::default()
+					}
+				}
+
+				/// Read current head pointer.
+				fn read_head<S: #scrate::GenericStorage>(storage: &S) -> Option<#kty> {
+					storage.get(#head_key.as_bytes())
+				}
+
+				/// Overwrite current head pointer.
+				///
+				/// If `None` is given head is removed from storage.
+				fn write_head<S: #scrate::GenericStorage>(storage: &S, head: Option<&#kty>) {
+					match head {
+						Some(head) => storage.put(#head_key.as_bytes(), head),
+						None => storage.kill(#head_key.as_bytes()),
+					}
+				}
+			}
+
+			#visibility struct #name<#traitinstance: #traittype>(#scrate::storage::generator::PhantomData<#traitinstance>);
+
+			impl<#traitinstance: #traittype> #scrate::storage::generator::StorageMap<#kty, #typ> for #name<#traitinstance> {
+				type Query = #value_type;
+
+				/// Get the prefix key in storage.
+				fn prefix() -> &'static [u8] {
+					#prefix.as_bytes()
+				}
+
+				/// Get the storage key used to fetch a value corresponding to a specific key.
+				fn key_for(x: &#kty) -> #scrate::rstd::vec::Vec<u8> {
+					let mut key = <Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::prefix().to_vec();
+					#scrate::codec::Encode::encode_to(x, &mut key);
+					key
+				}
+
+				/// Load the value associated with the given key from the map.
+				fn get<S: #scrate::GenericStorage>(key: &#kty, storage: &S) -> Self::Query {
+					storage.get(&*#key_for)
+						.#option_simple_1(|| #fielddefault)
+				}
+
+				/// Take the value, reading and removing it.
+				fn take<S: #scrate::GenericStorage>(key: &#kty, storage: &S) -> Self::Query {
+					let res: Option<(#value_type, #linkage)> = storage.take(&*#key_for);
+					match res {
+						Some((data, linkage)) => {
+							linkage.remove(storage, |key| #key_for);
+							data
+						},
+						None => #fielddefault
+					}
+				}
+
+				/// Remove the value under a key.
+				fn remove<S: #scrate::GenericStorage>(key: &#kty, storage: &S) {
+					<Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::take(key, storage);
+				}
+
+				/// Store a value to be associated with the given key from the map.
+				fn insert<S: #scrate::GenericStorage>(key: &#kty, val: &#typ, storage: &S) {
+					let key_for = &*#key_for;
+					let linkage = match #linkage::read(storage, key_for) {
+						// overwrite but reuse existing linkage
+						Some((_data, linkage)) => linkage,
+						// create new linkage
+						None => #linkage::insert_new_head(storage, key.clone(), |key| #key_for),
+					};
+					storage.put(key_for, &(*val, linkage))
```

Comment:
Do we need to clone the key for this `key_for` here? How does that work? Shouldn't `key_for` be an function pointer?